### PR TITLE
ArrayImageSource Patches for use in MeanEstimator

### LINKED ISSUE #426: Using ArrayImageSource With Mean/Volume Estimator Crashes
Found when trying to use researcher provided MATLAB reference data for #406 .

Essentially, `ArrayImageSource` does not correctly set attributes relating to rotations (`rots` property, and/ `_rotations`) required in some of the volume estimation code.  At least that is the first issue... Probably these just need to be initialized to zero equivalents. 

Reproduced in `develop`. The following isolated bit of code extends the unit tests to cover this case:

```
+++ b/tests/test_image.py
@@ -4,7 +4,9 @@ from unittest import TestCase
 import numpy as np
 from scipy import misc
 
+from aspire.basis import FBBasis3D
 from aspire.image import Image, _im_translate2
+from aspire.reconstruction import MeanEstimator
 from aspire.source import ArrayImageSource
 
 DATA_DIR = os.path.join(os.path.dirname(__file__), "saved_test_data")
@@ -12,13 +14,14 @@ DATA_DIR = os.path.join(os.path.dirname(__file__), "saved_test_data")
 
 class ImageTestCase(TestCase):
     def setUp(self):
+        self.dtype = np.float64
         # numpy array for top-level functions that directly expect it
-        self.im_np = misc.face(gray=True).astype(np.float64)[np.newaxis, :768, :768]
+        self.im_np = misc.face(gray=True).astype(self.dtype)[np.newaxis, :768, :768]
         # Independent Image object for testing Image methods
-        self.im = Image(misc.face(gray=True).astype(np.float64)[:768, :768])
+        self.im = Image(misc.face(gray=True).astype(self.dtype)[:768, :768])
         # Construct a simple stack of Images
         self.n = 3
-        self.ims_np = np.empty((3, *self.im_np.shape[1:]), dtype=self.im_np.dtype)
+        self.ims_np = np.empty((3, *self.im_np.shape[1:]), dtype=self.dtype)
         for i in range(self.n):
             self.ims_np[i] = self.im_np * (i + 1) / float(self.n)
         # Independent Image stack object for testing Image methods
@@ -53,6 +56,12 @@ class ImageTestCase(TestCase):
         im = src.images(start=0, num=np.inf)
         self.assertTrue(np.allclose(im.asnumpy(), self.im_np))
 
+    def testArrayImageSourceVol(self):
+        src = ArrayImageSource(self.im)
+        basis = FBBasis3D((8, 8, 8), dtype=self.dtype)
+        estimator = MeanEstimator(src, basis, preconditioner="none")
+        _ = estimator.estimate()
+
     def testImageSqrt(self):
         self.assertTrue(np.allclose(self.im.sqrt().asnumpy(), np.sqrt(self.im_np)))
 ```
 
 tb:
 
 ```
 ======================================================= FAILURES ========================================================
_________________________________________ ImageTestCase.testArrayImageSourceVol _________________________________________

self = <tests.test_image.ImageTestCase testMethod=testArrayImageSourceVol>

    def testArrayImageSourceVol(self):
        src = ArrayImageSource(self.im)
        basis = FBBasis3D((8, 8, 8), dtype=self.dtype)
        estimator = MeanEstimator(src, basis, preconditioner="none")
>       _ = estimator.estimate()

tests/test_image.py:63: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
src/aspire/reconstruction/estimator.py:68: in estimate
    b_coeff = self.src_backward()
src/aspire/reconstruction/estimator.py:85: in src_backward
    batch_mean_b = self.src.im_backward(im, i) / self.n
src/aspire/source/image.py:509: in im_backward
    vol = im.backproject(self.rots[start : start + num, :, :])[0]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <aspire.source.image.ArrayImageSource object at 0x11cea6160>

    @property
    def rots(self):
        """
        :return: Rotation matrices as a n x 3 x 3 array
        """
>       return self._rotations.as_matrix().astype(self.dtype)
E       AttributeError: 'ArrayImageSource' object has no attribute '_rotations'

src/aspire/source/image.py:195: AttributeError
 ```

### PR DESCRIPTION
**Pull request by garrettwrong**
Working through some issues actually trying to use `ArrayImageSource`. 

Adds optional `angles` arg to `ArrayImageSource`.
Breaks `ArrayImageSource` test into own file away from `ImageSource` (different size inputs).
Adds supporting tests covering the parameter and some misc things found in process of debugging.
Some light cleanup in areas I was poking around.

### CODE REVIEW

**Comment by garrettwrong** [comment]
I'll squash this after some review/discussion.

**Comment by garrettwrong** [comment]
Closes #426

File: src/aspire/source/image.py [hunk 1]
```diff
@@ -736,13 +760,59 @@ class ArrayImageSource(ImageSource):
...
+
+        # Optionally populate angles/rotations.
+        if angles is not None:
+            if angles.shape != (self.n, 3):
+                raise ValueError(f"Angles should be shape {(self.n, 3)}")
+            # This will populate `_rotations`,
+            #   which is exposed by properties `angles` and `rots`.
+            self.angles = angles
+
+    def _rots(self):
```
Comment (janden) [inline src/aspire/source/image.py:790]: Why not put this (and `_angles`) in the `ImageSource` class. Presumably, we'd want this type of behavior whenever rotations are missing.

**Review by janden — CHANGES_REQUESTED** [review]

File: src/aspire/source/image.py (hunk 1 above)
Comment (garrettwrong) [reply src/aspire/source/image.py:790]: I guess I was trying to balance the scope of the patch.

Yes, it does make sense for `ImageSource` and `ArrayImageSource`. However, I wasn't sure if it always made sense for any subclass. Specifically 2/3 that exist today, `mrcstack` and `RelionSource`.  To accommodate those would require refactoring the classes more.  It looks like it becomes writing a unique `_rots` and `_angles` for each of those remaining subclasses anyway, so it seems there is not much gained today by putting into `ImageSource` at this time.

FWIW, It's sort of older school python, but this is where Mix-Ins can be handy. I think they are frowned upon these days for some reasons... but favored approaches really require we get the details of the abstractions right; this wasn't done and is why I had to add the `_rots` and `_angles` methods... 

I think this is a solid patch covering stuff that was in my way.  Although my dark side thinks we should just eviscerate most of this code, I'm just trying to cover the ArrayImageSource case right now.  I am using this patch already to tinker with Amit H's data.... We can mimic or refactor more to extend further if and when these issues come up again.  Let's talk about it in the meeting tomorrow.

Comment (garrettwrong) [reply src/aspire/source/image.py:790]: I guess what I am saying is I think these source classes are just structured sub optimally. Fixing them all together requires refactoring/rewriting all of them, and is out of scope at this moment.

**Review by janden — APPROVED** [review]

**Pull request merged**